src/metabase/analytics/prometheus.clj

```diff
@@ -0,0 +1,235 @@
+      (throw (ex-info (trs "Failed to initialize Prometheus on port {0}" port)
+                      {:port port}
+                      e)))))
+
+;;; Collectors
+
+(defn c3p0-stats
+  "Takes `raw-stats` from [[metabase.troubleshooting/connection-pool-info]] and groups by each property type rather than each database.
+  {\"metabase-postgres-app-db\" {:numConnections 15,
+                                 :numIdleConnections 15,
+                                 :numBusyConnections 0,
+                                 :minPoolSize 1,
+                                 :maxPoolSize 15},
+   \"db-2-postgres-clean\" {:numConnections 2,
+                            :numIdleConnections 2,
+                            :numBusyConnections 0,
+                            :minPoolSize 1,
+                            :maxPoolSize 15}}
+  Becomes {:numConnections [{:name :numConnections,
+                             :value 15.0, ;; values are all doubles
+                             :timestamp 1662563931039,
+                             :label \"metabase-postgres-app-db\"}
+                            {:name :numConnections,
+                             :value 2.0,
+                             :timestamp 1662563931039,
+                             :label \"db-2-postgres-clean\"}]
+          ...}"
+  [raw-stats]
+  (let [now    (.toEpochMilli (java.time.Instant/now))
+        sample (fn sample [[db-label k v]]
+                 {:name      k
+                  :value     (double v)
+                  :timestamp now
+                  :label     db-label})]
+    (->> raw-stats
+         (mapcat (fn [[db-label values]]
+                   (map (fn [[k v]] [db-label k v]) values)))
+         (map sample)
+         (group-by :name))))
+
+(def ^:private label-translation
+  {:maxPoolSize        {:label       "c3p0_max_pool_size"
+                        :description (deferred-trs "C3P0 Max pool size")}
+   :minPoolSize        {:label       "c3p0_min_pool_size"
+                        :description (deferred-trs "C3P0 Minimum pool size")}
+   :numConnections     {:label       "c3p0_num_connections"
+                        :description (deferred-trs "C3P0 Number of connections")}
+   :numIdleConnections {:label       "c3p0_num_idle_connections"
+                        :description (deferred-trs "C3P0 Number of idle connections")}
+   :numBusyConnections {:label       "c3p0_num_busy_connections"
+                        :description (deferred-trs "C3P0 Number of busy connections")}})
+
+(defn- stats->prometheus
+  "Create an ArrayList of GaugeMetricFamily objects containing measurements from the c3p0 stats. Stats are grouped by
+  the property and the database information is attached as a label to multiple measurements of `:numConnections`."
+  [stats]
+  (let [arr (ArrayList. (count stats))]
+    (doseq [[raw-label measurements] stats
+            :let [{gauge-label :label desc :description} (label-translation raw-label)]
+            :when gauge-label
```

Comment:
Compressing this in a `when-let` would be even nicer, IMHO.

Reply:
using an `if-let` made it simpler and gave a spot to bring the `log` back in. Good thinking.